{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import math\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import csv\n",
    "import os\n",
    "\n",
    "from utils_ff import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Define parameters used in script:\n",
    "#Reaction parameters\n",
    "v_plasma = 14.5 #in ml (volume used in plasma extraction)\n",
    "v_elution = 150. #in ul (volume of purified DNA obtained from v_plasma)\n",
    "v_ddpcr = 20 #in ul (standard volume of ddPCR)\n",
    "droplet_size = 0.85e-3 #in (approx value from BioRad ddPCR)\n",
    "v_input_quant = 1 #in ul (input DNA from quantification, for 20ul total reaction)\n",
    "v_input_preamp = 8.5 #in ul (input DNA for preamp)\n",
    "\n",
    "#Analysis and plot parameters\n",
    "#Threshold values of MAF to pick SNPs used for fetral fraction quantification \n",
    "min_thres=.75 # Minimum MAF threshold to pick SNP for fetal fraction determination (MAF = fetal_frac / 2)\n",
    "max_thres=20. # Max MAF threshold to pick SNP for fetal fraction determination\n",
    "SNP_min_droplets = 100 #Min number of positive droplets (FAM+VIC) to use assay in fetal fraction calculation\n",
    "resolfrac=0.5 #Resolution of histogram (in % MAF)\n",
    "maxfrac=50 #Maximum value of histogram (in % MAF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams[\"font.family\"] = 'Sans-serif'\n",
    "matplotlib.rcParams[\"font.size\"] = 8\n",
    "\n",
    "input_folder = './NIPT_FF/'\n",
    "csv_folder = 'csv_files/'\n",
    "input_file = input_folder + csv_folder + '20170524_NIPT017FF.csv'\n",
    "Sample_name = 'NIPT017'\n",
    "\n",
    "SNP_assays = ('SNP1-FAM','SNP2-FAM','SNP3-FAM','SNP4-FAM','SNP5-FAM',\n",
    "             'SNP7-FAM','SNP9-FAM','SNP10-FAM','SNP12-FAM','SNP13-FAM',\n",
    "             'SNP14-FAM','SNP16-FAM','SNP17-FAM','SNP18-FAM','SNP19-FAM',\n",
    "            'SNP20-FAM','SNP22-FAM','SNP23-FAM','SNP24-FAM','SNP25-FAM',\n",
    "            'SNP26-FAM','SNP27-FAM','SNP28-FAM','SNP30-FAM','SNP31-FAM',\n",
    "            'SNP32-FAM','SNP33-FAM','SNP34-FAM','SNP35-FAM','SNP36-FAM',\n",
    "            'SNP37-FAM','SNP38-FAM','SNP39-FAM','SNP42-FAM',\n",
    "            'SNP44-FAM','SNP46-FAM','SNP47-FAM','SNP48-FAM','SNP49-FAM',\n",
    "            'SNP50-FAM','SNP51-FAM','SNP52-FAM','SNP55-FAM','SNP56-FAM',\n",
    "            'SNP57-FAM','SNP58-FAM','SNP59-FAM','Y1-FAM')\n",
    "Quant_assays = ('EIF2C1-FAM','RPP30-FAM','TERT-VIC')\n",
    "#'RNaseP-VIC', remove this quant assay as it underestimates\n",
    "list_na =['No Call']\n",
    "\n",
    "cols_snp=['Sample','Target','Conc(copies/µL)','Positives','Negatives',\n",
    "          'Ch1+Ch2+', 'Ch1+Ch2-', 'Ch1-Ch2+', 'Ch1-Ch2-']\n",
    "\n",
    "cols_quant=['Sample','Target','Conc(copies/µL)','Positives','Negatives',\n",
    "            'PoissonConfidenceMax68', 'PoissonConfidenceMin68']\n",
    "\n",
    "log_columns = ['Parameter', Sample_name]\n",
    "\n",
    "\n",
    "output_folder = input_folder + Sample_name + '/output_files/'\n",
    "figure_folder = input_folder + Sample_name + '/Figures/'\n",
    "log_folder = input_folder + Sample_name + '/log_files/'\n",
    "\n",
    "if not os.path.exists(output_folder):\n",
    "    os.makedirs(output_folder)\n",
    "    \n",
    "if not os.path.exists(figure_folder):\n",
    "    os.makedirs(figure_folder)\n",
    "\n",
    "if not os.path.exists(log_folder):\n",
    "    os.makedirs(log_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def CreateMAFDataframe(df, Sample_name, SNP_assays, SNP_min_droplets, cols_snp):\n",
    "    '''\n",
    "    Calculate Minor Allele Fraction of SNP assay panel.\n",
    "    First filter df for columns of interest and removing SNP assays with low droplet counts\n",
    "    df: Dataframe from Biorad ddPCR experiment\n",
    "    SNP_assay: list with SNP assay names (Target column)\n",
    "    SNP_min_droplets: Minimum total number of droplets (FAM+VIC) to consider assat useful\n",
    "    cols_snp: columns required for snp analyisis.\n",
    "    '''    \n",
    "    try:\n",
    "        rows_filter = ((df.Target.isin(SNP_assays)) &\n",
    "                       (df.Sample == Sample_name) &\n",
    "                       (df['Ch1+Ch2+']+df['Ch1+Ch2-']+df['Ch1-Ch2+'] > SNP_min_droplets))\n",
    "        \n",
    "        new_df = df.loc[rows_filter, cols_snp]\n",
    "        \n",
    "        N_total = new_df['Ch1+Ch2+'] + new_df['Ch1-Ch2+'] + new_df['Ch1+Ch2-'] + new_df['Ch1-Ch2-']\n",
    "        N_allele1 = new_df['Ch1+Ch2+'] + new_df['Ch1+Ch2-']\n",
    "        N_allele2 = new_df['Ch1+Ch2+'] + new_df['Ch1-Ch2+']\n",
    "\n",
    "        new_df['Allele 1'] = - np.log(1.0 - N_allele1 / N_total) * N_total\n",
    "        new_df['Allele 2'] = - np.log(1.0 - N_allele2 / N_total) * N_total\n",
    "        new_df['MAF'] =  np.abs( new_df[['Allele 1','Allele 2']].min(axis=1) / \n",
    "                                (new_df['Allele 1'] + new_df['Allele 2']) * 100. )\n",
    "\n",
    "    except IOError as e:\n",
    "        errno, strerror = e.args\n",
    "        print(\"I/O error({0}): {1}\".format(errno, strerror))    \n",
    "    \n",
    "    except ZeroDivisionError as err:\n",
    "        print('Handling run-time error:', err)\n",
    "    \n",
    "    except ValueError:\n",
    "        print(\"No valid integer in line.\")\n",
    "         \n",
    "    return new_df\n",
    "\n",
    "\n",
    "def CreateQuantDataframe(df, Sample_name, Quant_assays, cols_quant):\n",
    "    '''\n",
    "    Create Dataframe with cfDNA quantification panel.\n",
    "    '''    \n",
    "    rows_filter = ((df.Target.isin(Quant_assays)) \n",
    "                   & (df.Sample == Sample_name)\n",
    "                   & (df.Positives != 0))\n",
    "\n",
    "    new_df = df.loc[rows_filter, cols_quant]       \n",
    "    new_df['SD_Max_Qsoft'] = new_df['PoissonConfidenceMax68'] - new_df['Conc(copies/µL)']\n",
    "    new_df['SD_Min_Qsoft'] = new_df['Conc(copies/µL)'] - new_df['PoissonConfidenceMin68']\n",
    "\n",
    "    N_total = new_df['Positives'] + new_df['Negatives']\n",
    "    N_positive = new_df['Positives']\n",
    "\n",
    "    new_df['total_DNA_input'] = - (np.log(1.0- N_positive / N_total) * N_total \n",
    "                                /  (droplet_size * N_total) * v_ddpcr)\n",
    "\n",
    "    #Error propagation of counts\n",
    "    new_df['SD_total_DNA_input'] = (np.sqrt(N_positive) / (N_total-N_positive) * N_total / \n",
    "                                    (droplet_size * N_total) * v_ddpcr)\n",
    "    #Compare errors quantasoft\n",
    "    new_df['errors_quantasoft'] = v_ddpcr * ((new_df['SD_Max_Qsoft'] + new_df['SD_Min_Qsoft']) / 2.)\n",
    "    \n",
    "    return new_df\n",
    "\n",
    "def print_results(df_quant, df_snp):\n",
    "\n",
    "    print(\"Parameters used:\\n\\\n",
    "    --------------------------------------------------\\n\\\n",
    "    volume plasma extracted: %.1f ml volume elution DNA: %.1f ul\\n\\\n",
    "    volume input DNA quantification: %.1f ul\\n\\n\\\n",
    "    volume ddpcr reaction: %.1f ul, droplet_size: %.1f nl\\n\\n\"\n",
    "            % (v_plasma,v_elution,v_input_quant,v_ddpcr, droplet_size*1000,))\n",
    "\n",
    "    print(\"DNA input for quantification (in genomic copies):\\n\\\n",
    "    --------------------------------------------------\\n\\\n",
    "    Mean: %.1f, Median: %.1f. SD:%.1f\\nAverage Poisson noise: %.1f \\\n",
    "    Quantasoft Poisson noise: %.1f\\n\\n\"\n",
    "            % (df_quant.total_DNA_input.mean(), df_quant.total_DNA_input.median(), \n",
    "              df_quant.total_DNA_input.std(), df_quant.SD_total_DNA_input.mean(), df_quant.errors_quantasoft.mean()))\n",
    "\n",
    "\n",
    "    #Output of plasma in 1ml basis\n",
    "    print(\"Concentration cfDNA per ml plasma (in genomic copies):\\n\\\n",
    "    ------------------------------------------------------\\n\\\n",
    "    Total: %.1f+-%.1f (%.1f) [mean+-SD (Poisson SD)]\" % (C_cfDNA, SD_cfDNA, PoissSD_cfDNA))\n",
    "    \n",
    "    print(\"Number of SNPs in range MAF: %i-%i%%: %i\\n\\\n",
    "    \\tMean = %4.2f\\n\\tMedian = %4.2f\\n\\tSD = %4.2f\\n\\\n",
    "    \\tY chromosome fraction: %.1f%%\\n\\n\\\n",
    "    Expected SD from quantitation error: %4.2f\\n\\\n",
    "    Expected SD from fetal allele noise: %4.2f\"\n",
    "    %( min_thres, max_thres, len(ffrac), ffrac.mean(), ffrac.median(),\n",
    "      ffrac.std(), Y_fraction, Err_fetfrac_1, Err_fetfrac_2))\n",
    "\n",
    "def Make_SNP_plot():\n",
    "  \n",
    "    matplotlib.rcParams['xtick.labelsize'] = 16 \n",
    "    matplotlib.rcParams['ytick.labelsize'] = 16 \n",
    "    matplotlib.rcParams['axes.labelsize'] = 16 \n",
    "    hist, bin_edges = np.histogram(df_snp['MAF'],bins=maxfrac/resolfrac,range=(0,maxfrac))\n",
    "    cumhist = np.cumsum(hist)\n",
    "\n",
    "    #Historam figure\n",
    "    fig, ax = plt.subplots(1,2,figsize=(20,5))\n",
    "    fig.suptitle(Sample_name, fontsize = 18 )\n",
    "    x = np.arange(min(bin_edges), max(bin_edges), 0.1);\n",
    "    y = len(ffrac)/(math.sqrt(2*3.14*(ffrac.std()/2)**2)) * np.exp(-(x-ffrac.median()/2)**2/(2*(ffrac.std()/2)**2))\n",
    "    ax[0].plot(x, y)\n",
    "    ax[0].bar(bin_edges[:-1],hist,width=resolfrac,color='#7694C5')\n",
    "    ax[0].set_ylim(0,max(hist)+1)\n",
    "    ax[0].set_xlim(min(bin_edges), max(bin_edges))\n",
    "    ax[0].axvline(x=min_thres, ymin=min(bin_edges), ymax=max(bin_edges), linewidth=0.5, color ='k')\n",
    "    ax[0].axvline(x=max_thres, ymin=min(bin_edges), ymax=max(bin_edges), linewidth=0.5, color ='k')\n",
    "    ax[0].set_xlabel('Minor Allele fraction observed (%)')\n",
    "\n",
    "    ax[0].set_ylabel('# of SNP tests')\n",
    "    #Plot text squares\n",
    "    ax[0].text(0.95,0.95, 'N= %i, Fetal fraction: %3.1f$\\pm$%3.1f%% ' % (ffrac.len(), ffrac.median(),ffrac.std()),\n",
    "            style='italic', transform=ax[0].transAxes, horizontalalignment='right', verticalalignment='top',\n",
    "                fontsize=18, bbox={'facecolor':'white', 'alpha':0.8, 'pad':8})\n",
    "    ax[0].text(0.95,0.8, 'Y chromosome: %.1f%% ' % (Y_fraction),\n",
    "            style='italic', transform=ax[0].transAxes, horizontalalignment='right', verticalalignment='top',\n",
    "                fontsize=18, bbox={'facecolor':'white', 'alpha':0.8, 'pad':8})\n",
    "\n",
    "    #Cumulative sum subplot\n",
    "    ax[1].plot(bin_edges[:-1], cumhist)\n",
    "    plt.xlim(min(bin_edges), max(bin_edges))\n",
    "    ax[1].set_ylim(0,max(cumhist)+1)\n",
    "    ax[1].set_xlim(min(bin_edges), max(bin_edges))\n",
    "    ax[1].set_xlabel('Minor Allele fraction observed (%)')\n",
    "    ax[1].set_ylabel('Cumulative of SNP tests')\n",
    "    ax[1].text(0.05,0.95, \"[cfDNA] = %.1f$\\pm$%.1f (%.1f) genomic copies / ml plasma\"\n",
    "               % (C_cfDNA, SD_cfDNA, PoissSD_cfDNA), style='italic', transform=ax[1].transAxes, \n",
    "                   horizontalalignment='left', verticalalignment='top',\n",
    "                fontsize=16, bbox={'facecolor':'white', 'alpha':0.8, 'pad':8})\n",
    "    ax[1].text(0.05,0.8, \"Expected SD in fetal fraction: %.1f%% (%.1f%%)\"\n",
    "               % (Err_fetfrac_1, Err_fetfrac_2), style='italic', transform=ax[1].transAxes, \n",
    "               horizontalalignment='left', verticalalignment='top',\n",
    "                fontsize=16, bbox={'facecolor':'white', 'alpha':0.8, 'pad':8})\n",
    "\n",
    "\n",
    "    #Plot inset\n",
    "    left, bottom, width, height = [0.26, 0.4, 0.2, 0.25]\n",
    "    axinset = fig.add_axes([left, bottom, width, height])\n",
    "    axinset.set_axis_bgcolor((1, 1, 1))\n",
    "    axinset.plot(x, y)\n",
    "    axinset.bar(bin_edges[:-1],hist,width=resolfrac,color='#7694C5')\n",
    "    axinset.set_ylim(0,10)\n",
    "    axinset.set_xlim(min(bin_edges), ffrac.mean())\n",
    "    axinset.tick_params(axis='both', labelsize=14) \n",
    "    fig.tight_layout()\n",
    "    fig.savefig( figure_folder + Sample_name + '_fig.pdf')\n",
    "    \n",
    "def QuantifyDNAInput(df_quant):\n",
    "    C_cfDNA =  df_quant.total_DNA_input.mean() * v_elution / v_input_quant / v_plasma\n",
    "    SD_cfDNA =  df_quant.total_DNA_input.std() * v_elution / v_input_quant / v_plasma\n",
    "    PoissSD_cfDNA =  df_quant.SD_total_DNA_input.mean() * v_elution / v_input_quant / v_plasma\n",
    "    DNA_input_preamp = C_cfDNA * v_plasma * (v_input_preamp / v_elution)\n",
    "    Err_input_preamp = SD_cfDNA * v_plasma * (v_input_preamp / v_elution)\n",
    "    return C_cfDNA, SD_cfDNA, PoissSD_cfDNA, DNA_input_preamp, Err_input_preamp\n",
    "    \n",
    "\n",
    "def QuantifySNPresults(df_snp):\n",
    "    #Determine fetal fraction with all alleles between 1-20%\n",
    "    ffrac = 2 * df_snp['MAF'][(df_snp['MAF']>min_thres) & (df_snp['MAF']<max_thres)]\n",
    "\n",
    "    #Expected SD in fetal fraction due to fluctuations of input preAmp DNA (assuming perfect amplification)\n",
    "    #First determine expected counts coming from baby\n",
    "    InputDNA_Fetal_Allele = DNA_input_preamp * ( ffrac.median() / 2 ) / 100\n",
    "    Err_InputDNA_Fetal_Allele  = Err_input_preamp * ( ffrac.median() / 2 ) / 100 \n",
    "    ErrPoiss_InputDNA_Fetal_Allele  = math.sqrt( InputDNA_Fetal_Allele )\n",
    "\n",
    "    #Calculate Errors in Fetal Fraction\n",
    "    Err_fetfrac_1 = ( Err_InputDNA_Fetal_Allele / DNA_input_preamp ) * 100 * 2\n",
    "    Err_fetfrac_2 = ( ErrPoiss_InputDNA_Fetal_Allele / DNA_input_preamp ) * 100 * 2\n",
    "\n",
    "    #Calculate fraction of Y Chromosome\n",
    "    Y_fraction = abs(df_snp[df_snp.Target == 'Y1-FAM']['MAF'].values[0])\n",
    "    return ffrac, InputDNA_Fetal_Allele, Err_InputDNA_Fetal_Allele, \\\n",
    "    ErrPoiss_InputDNA_Fetal_Allele, Err_fetfrac_1, Err_fetfrac_2, Y_fraction\n",
    "    \n",
    "def WriteListToCSV(csv_file,csv_columns,data_list):\n",
    "    try:\n",
    "        with open(csv_file, 'w') as csvfile:\n",
    "            writer = csv.writer(csvfile)\n",
    "            writer.writerow(csv_columns)\n",
    "            for data in data_list:\n",
    "                writer.writerow(data)\n",
    "    except IOError as e:\n",
    "        errno, strerror = e.args\n",
    "        print(\"I/O error({0}): {1}\".format(errno, strerror))    \n",
    "    \n",
    "    except ValueError:\n",
    "        print(\"No valid integer in line.\")\n",
    "        \n",
    "    return\n",
    "\n",
    "def MakeOutputTable(df_quant, df_snp):\n",
    "    #Prepare SNP results for output (output minor allele fraction of each SNP)\n",
    "    df1 = df_snp[['Target','MAF']].copy()\n",
    "    df1['Target'] = df1.Target.map(lambda x: x.rstrip('-FAMVIC'))\n",
    "    df1['#'] = [int(snp[3:]) if snp[:3] == 'SNP' else int(1e10) for snp in df1['Target'].values]\n",
    "    df1.sort_values(by='#',inplace=True)\n",
    "    df1.drop('#', inplace=True, axis=1)\n",
    "    df1 = df1.set_index('Target')\n",
    "    df1.columns = [ Sample_name ]\n",
    "\n",
    "    #Prepare quantification results for output (output total DNA input used in quantification)\n",
    "    df2 = df_quant[['Target','total_DNA_input']].copy()#output_table\n",
    "    df2 = df2.set_index('Target')\n",
    "    df2 = df2.rename({'EIF2C1-FAM': 'Q1', 'RNaseP-VIC': 'Q2', 'RPP30-FAM' : 'Q3', 'TERT-VIC' : 'Q4'})\n",
    "    df2.sort_index(inplace=True)\n",
    "    df2.columns = [ Sample_name ]\n",
    "\n",
    "    #Merge output information and save in folder\n",
    "    df_merge = pd.concat((df1,df2))\n",
    "    return df_merge\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def Make_SNP_plot2():\n",
    "  \n",
    "    matplotlib.rcParams['xtick.labelsize'] = 16 \n",
    "    matplotlib.rcParams['ytick.labelsize'] = 16 \n",
    "    matplotlib.rcParams['axes.labelsize'] = 16 \n",
    "    hist, bin_edges = np.histogram(df_snp['MAF'],bins=maxfrac/resolfrac,range=(0,maxfrac))\n",
    "    cumhist = np.cumsum(hist)\n",
    "\n",
    "    #Historam figure\n",
    "    fig, ax = plt.subplots(1,1,figsize=(10,5))\n",
    "    fig.suptitle(Sample_name, fontsize = 18 )\n",
    "    x = np.arange(min(bin_edges), max(bin_edges), 0.1);\n",
    "    y = len(ffrac)/(math.sqrt(2*3.14*(ffrac.std()/2)**2)) * np.exp(-(x-ffrac.median()/2)**2/(2*(ffrac.std()/2)**2))\n",
    "    ax.plot(x, y)\n",
    "    ax.bar(bin_edges[:-1],hist,width=resolfrac,color='#7694C5')\n",
    "    ax.set_ylim(0,max(hist)+1)\n",
    "    ax.set_xlim(min(bin_edges), max(bin_edges))\n",
    "    ax.axvline(x=min_thres, ymin=min(bin_edges), ymax=max(bin_edges), linewidth=0.5, color ='k')\n",
    "    ax.axvline(x=max_thres, ymin=min(bin_edges), ymax=max(bin_edges), linewidth=0.5, color ='k')\n",
    "    ax.set_xlabel('Minor Allele fraction observed (%)')\n",
    "\n",
    "    ax.set_ylabel('# of SNP tests')\n",
    "    #Plot text squares\n",
    "    ax.text(0.95,0.95, 'N= %i, Fetal fraction: %3.1f$\\pm$%3.1f%% ' % (len(ffrac), ffrac.median(),ffrac.std()),\n",
    "            style='italic', transform=ax.transAxes, horizontalalignment='right', verticalalignment='top',\n",
    "                fontsize=18, bbox={'facecolor':'white', 'alpha':0.8, 'pad':8})\n",
    "    ax.text(0.95,0.8, 'Y chromosome: %.1f%% ' % (Y_fraction),\n",
    "            style='italic', transform=ax.transAxes, horizontalalignment='right', verticalalignment='top',\n",
    "                fontsize=18, bbox={'facecolor':'white', 'alpha':0.8, 'pad':8})\n",
    "\n",
    "    ax.text(0.03,0.95, \"[cfDNA] = %.1f$\\pm$%.1f gen./ml plasma\"\n",
    "               % (C_cfDNA, SD_cfDNA), style='italic', transform=ax.transAxes, \n",
    "                   horizontalalignment='left', verticalalignment='top',\n",
    "                fontsize=16, bbox={'facecolor':'white', 'alpha':0.8, 'pad':8})\n",
    "    ax.text(0.03,0.82, \"Expected SD in FF: %.1f%%\"\n",
    "               % (Err_fetfrac_2), style='italic', transform=ax.transAxes, \n",
    "               horizontalalignment='left', verticalalignment='top',\n",
    "                fontsize=16, bbox={'facecolor':'white', 'alpha':0.8, 'pad':8})\n",
    "\n",
    "    #Plot inset\n",
    "    left, bottom, width, height = [0.55, 0.45, 0.4, 0.25]\n",
    "    axinset = fig.add_axes([left, bottom, width, height])\n",
    "    axinset.set_axis_bgcolor((1, 1, 1))\n",
    "    axinset.plot(x, y)\n",
    "    axinset.bar(bin_edges[:-1],hist,width=resolfrac,color='#7694C5')\n",
    "    axinset.set_ylim(0,10)\n",
    "    axinset.set_xlim(min(bin_edges), ffrac.mean())\n",
    "    axinset.tick_params(axis='both', labelsize=14) \n",
    "    fig.tight_layout()\n",
    "    fig.savefig( figure_folder + Sample_name + '_fig.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "//anaconda/lib/python3.5/site-packages/numpy/lib/function_base.py:564: VisibleDeprecationWarning: using a non-integer number instead of an integer will result in an error in the future\n",
      "  n = np.zeros(bins, ntype)\n",
      "//anaconda/lib/python3.5/site-packages/numpy/lib/function_base.py:600: VisibleDeprecationWarning: using a non-integer number instead of an integer will result in an error in the future\n",
      "  n += np.bincount(indices, weights=tmp_w, minlength=bins).astype(ntype)\n",
      "//anaconda/lib/python3.5/site-packages/matplotlib/figure.py:1744: UserWarning: This figure includes Axes that are not compatible with tight_layout, so its results might be incorrect.\n",
      "  warnings.warn(\"This figure includes Axes that are not \"\n",
      "//anaconda/lib/python3.5/site-packages/matplotlib/tight_layout.py:222: UserWarning: tight_layout : falling back to Agg renderer\n",
      "  warnings.warn(\"tight_layout : falling back to Agg renderer\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parameters used:\n",
      "    --------------------------------------------------\n",
      "    volume plasma extracted: 14.5 ml volume elution DNA: 150.0 ul\n",
      "    volume input DNA quantification: 1.0 ul\n",
      "\n",
      "    volume ddpcr reaction: 20.0 ul, droplet_size: 0.8 nl\n",
      "\n",
      "\n",
      "DNA input for quantification (in genomic copies):\n",
      "    --------------------------------------------------\n",
      "    Mean: 101.8, Median: 103.1. SD:20.9\n",
      "Average Poisson noise: 12.8     Quantasoft Poisson noise: 12.8\n",
      "\n",
      "\n",
      "Concentration cfDNA per ml plasma (in genomic copies):\n",
      "    ------------------------------------------------------\n",
      "    Total: 1052.6+-216.4 (132.5) [mean+-SD (Poisson SD)]\n",
      "Number of SNPs in range MAF: 0-20%: 13\n",
      "    \tMean = 11.18\n",
      "\tMedian = 11.20\n",
      "\tSD = 2.75\n",
      "    \tY chromosome fraction: 0.0%\n",
      "\n",
      "    Expected SD from quantitation error: 2.30\n",
      "    Expected SD from fetal allele noise: 1.61\n"
     ]
    }
   ],
   "source": [
    "data = pd.read_csv(input_file, sep=',', index_col=False,usecols=cols_snp + cols_quant ,na_values=list_na)\n",
    "\n",
    "df_snp = CreateMAFDataframe(data, Sample_name, SNP_assays, SNP_min_droplets, cols_snp)\n",
    "\n",
    "df_quant = CreateQuantDataframe(data, Sample_name, Quant_assays, cols_quant)\n",
    "\n",
    "#Quantification calculations\n",
    "C_cfDNA, SD_cfDNA, PoissSD_cfDNA, DNA_input_preamp, Err_input_preamp =  QuantifyDNAInput(df_quant)\n",
    "\n",
    "#Quantifications from SNP results\n",
    "ffrac, InputDNA_Fetal_Allele, Err_InputDNA_Fetal_Allele, ErrPoiss_InputDNA_Fetal_Allele,\\\n",
    "Err_fetfrac_1, Err_fetfrac_2, Y_fraction = QuantifySNPresults(df_snp)\n",
    "\n",
    "#Make plot\n",
    "Make_SNP_plot2()\n",
    "\n",
    "#Make output table\n",
    "output_table = MakeOutputTable(df_quant, df_snp)\n",
    "\n",
    "output_table.to_csv( output_folder + Sample_name + '_MAF.csv')\n",
    "\n",
    "#Print log file with main results\n",
    "log_list = [['v_plasma_ml', v_plasma], \n",
    "            ['v_elution_ul', v_elution], \n",
    "            ['v_ddpcr_ul', v_ddpcr],\n",
    "            ['droplet_size_nl', droplet_size*1000.],\n",
    "            ['v_input_quant_ul', v_input_quant],\n",
    "            ['v_input_preamp_ul', v_input_preamp],\n",
    "            ['MAF_min_thres', min_thres], \n",
    "            ['MAF_max_thres', max_thres],\n",
    "            ['Mean_DNA_Quantif_counts', df_quant.total_DNA_input.mean()],\n",
    "            ['SD_DNA_Quantif_counts', df_quant.total_DNA_input.std()], \n",
    "            ['SD_Poisson_counts', df_quant.SD_total_DNA_input.mean()], \n",
    "            ['Mean_cfDNA_counts_per_ml_plasma', C_cfDNA],\n",
    "            ['SD_cfDNA_counts_per_ml_plasma', SD_cfDNA],\n",
    "            ['SD_Poisson_cfDNA_counts_per_ml_plasma', PoissSD_cfDNA],\n",
    "            ['Min_range_ffrac', min_thres],\n",
    "            ['Max_range_ffrac', max_thres],\n",
    "            ['SNPs_in_ffrac_range', len(ffrac)],\n",
    "            ['Fetal_fraction_Mean', ffrac.mean()],\n",
    "            ['Fetal_fraction_Median', ffrac.median()],\n",
    "            ['Fetal_fraction_SD', ffrac.std()],\n",
    "            ['Y_chromosome_fraction', Y_fraction],\n",
    "            ['Expected_SD_from_quant_error', Err_fetfrac_1],\n",
    "            ['Expected_SD_from_Poisson_MAF', Err_fetfrac_2]]\n",
    "\n",
    "WriteListToCSV(log_folder + Sample_name + '_log.csv',log_columns,log_list)\n",
    "\n",
    "#Print main results on screen\n",
    "print_results(df_quant, df_snp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "log_list = [['v_plasma_ml', v_plasma], \n",
    "            ['v_elution_ul', v_elution], \n",
    "            ['v_ddpcr_ul', v_ddpcr],\n",
    "            ['droplet_size_nl', droplet_size*1000.],\n",
    "            ['v_input_quant_ul', v_input_quant],\n",
    "            ['v_input_preamp_ul', v_input_preamp],\n",
    "            ['MAF_min_thres', min_thres], \n",
    "            ['MAF_max_thres', max_thres],\n",
    "            ['Mean_DNA_Quantif_counts', df_quant.total_DNA_input.mean()],\n",
    "            ['SD_DNA_Quantif_counts', df_quant.total_DNA_input.std()], \n",
    "            ['SD_Poisson_counts', df_quant.SD_total_DNA_input.mean()], \n",
    "            ['Mean_cfDNA_counts_per_ml_plasma', C_cfDNA],\n",
    "            ['SD_cfDNA_counts_per_ml_plasma', SD_cfDNA],\n",
    "            ['SD_Poisson_cfDNA_counts_per_ml_plasma', PoissSD_cfDNA],\n",
    "            ['Min_range_ffrac', min_thres],\n",
    "            ['Max_range_ffrac', max_thres],\n",
    "            ['SNPs_in_ffrac_range', len(ffrac)],\n",
    "            ['Fetal_fraction_Mean', ffrac.mean()],\n",
    "            ['Fetal_fraction_Median', ffrac.median()],\n",
    "            ['Fetal_fraction_SD', ffrac.std()],\n",
    "            ['Y_chromosome_fraction', Y_fraction],\n",
    "            ['Expected_SD_from_quant_error', Err_fetfrac_1],\n",
    "            ['Expected_SD_from_Poisson_MAF', Err_fetfrac_2]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
